File: ScriptsToProcess/Fillet_Rules.py
"""
FILLET RULE CONFIGURATOR
================================================================================
A UI helper node that compiles standard Grasshopper lists into a strict JSON 
schema for parametric corner rounding. Prevents syntax errors from manual typing.

HOW TO USE NEGATIVE RULES (e.g., "Everything except Office"):
Because the Engine reads top-down, put your exception first, then a wildcard:
1. Type: Program | Match: Office | Radius: 0.0 (Exception)
2. Type: Program | Match: * | Radius: 3.0 (Everything Else)

INPUTS:
    DefaultRadius (float) : The base radius if no rules match.
    RuleTypes     (str)   : List of target types ('Tower', 'Program', 'Building').
    RuleMatches   (str)   : List of target names ('Main_Tower', 'Retail', '*').
    RuleRadii     (float) : List of radii to apply to the matched targets.
    ExactMatch    (bool)  [List Access] : True = Exact, False = Contains. Defaults to True.

OUTPUTS:
    Fillet_JSON   (str)   : The compiled JSON string to feed into BIM_JSON.
    Instructions  (str)   : Detailed usage instructions.
================================================================================
"""

import json

def update_ui(msg):
    _ghenv = globals().get('ghenv')
    _comp = globals().get('component')
    if _ghenv: _ghenv.Component.Message = msg
    elif _comp: _comp.Message = msg

_ghenv = globals().get('ghenv')
if _ghenv:
    _ghenv.Component.Name = "Fillet Rule Configurator"
    _ghenv.Component.NickName = "Fillet_Rules"
    _ghenv.Component.Description = "Compiles Grasshopper lists into a Fillet Rules JSON."

def build_config():
    # 1. Grab inputs
    default_rad = globals().get('DefaultRadius')
    try: default_rad = float(default_rad)
    except: default_rad = 0.0
        
    raw_types = globals().get('RuleTypes') or []
    raw_matches = globals().get('RuleMatches') or []
    raw_radii = globals().get('RuleRadii') or []
    
    # Grab ExactMatch as a List
    raw_exacts = globals().get('ExactMatch')
    if raw_exacts is None: raw_exacts = []
    
    # 2. Clean data
    types = [str(t).strip() for t in raw_types if t is not None]
    matches = [str(m).strip() for m in raw_matches if m is not None]
    radii = []
    for r in raw_radii:
        try: radii.append(float(r))
        except: radii.append(0.0)
            
    rules = []
    
    # 3. Zip lists together securely
    rule_count = min(len(types), len(matches), len(radii))
    
    for i in range(rule_count):
        # Safely get the exact match boolean for this index, defaulting to True
        is_exact = True
        if i < len(raw_exacts) and raw_exacts[i] is not None:
            is_exact = bool(raw_exacts[i])
            
        rules.append({
            "type": types[i],
            "match": matches[i],
            "radius": radii[i],
            "exact": is_exact
        })
        
    # 4. Construct JSON dictionary
    config_dict = {
        "default_radius": default_rad,
        "rules": rules
    }
    
    update_ui(f"RULES BUILT\n---\nValid Rules: {rule_count}\nDefault: {default_rad}m")
    
    # 5. Output formatted JSON string
    return json.dumps(config_dict, indent=2)

# Outputs
Fillet_JSON = build_config()
Instructions = __doc__

File: Components/FilletRules.cs
using System;
using System.Collections.Generic;
using Grasshopper.Kernel;
using Newtonsoft.Json;

namespace Enzyme.Components
{
    public class FilletRulesComponent : GH_Component
    {
        public FilletRulesComponent()
          : base("Fillet Rule Configurator", "Fillet_Rules",
              "Compiles Grasshopper lists into a Fillet Rules JSON.",
              "Enzyme", "Masterplan (Beta)")
        {
        }

                public override void AddedToDocument(GH_Document document)
        {
            base.AddedToDocument(document);
            if (this.Attributes == null) this.CreateAttributes();

            bool hasSources = false;
            foreach (var param in this.Params.Input)
                if (param.SourceCount > 0) { hasSources = true; break; }

            if (!hasSources)
            {
                Enzyme.Utils.AutoWireHelper.WireSlider(this, document, 0, 0.0, 2.0, 0.0, 330, -40);
                Enzyme.Utils.AutoWireHelper.WireValueList(this, document, 1, new string[]{"Tower", "Program", "Building"}, new string[]{"\"Tower\"", "\"Program\"", "\"Building\""}, 300, 0);
                Enzyme.Utils.AutoWireHelper.WireValueList(this, document, 2, new string[]{"Main_Tower", "Retail", "*"}, new string[]{"\"Main_Tower\"", "\"Retail\"", "\"*\""}, 300, 40);
                Enzyme.Utils.AutoWireHelper.WireOutputPanel(this, document, 0, 220, -34, 180, 22);
                Enzyme.Utils.AutoWireHelper.WireOutputPanel(this, document, 1, 220, 11, 180, 22);
            }
        }

        protected override void RegisterInputParams(GH_InputParamManager pManager)
        {
            pManager.AddNumberParameter("DefaultRadius", "DefaultRadius", "The base radius if no rules match.", GH_ParamAccess.item, 0.0);
            pManager.AddTextParameter("RuleTypes", "RuleTypes", "List of target types ('Tower', 'Program', 'Building').", GH_ParamAccess.list);
            pManager.AddTextParameter("RuleMatches", "RuleMatches", "List of target names ('Main_Tower', 'Retail', '*').", GH_ParamAccess.list);
            pManager.AddNumberParameter("RuleRadii", "RuleRadii", "List of radii to apply to the matched targets.", GH_ParamAccess.list);
            pManager.AddBooleanParameter("ExactMatch", "ExactMatch", "True = Exact, False = Contains. Defaults to True.", GH_ParamAccess.list);
            
            pManager[1].Optional = true;
            pManager[2].Optional = true;
            pManager[3].Optional = true;
            pManager[4].Optional = true;
        }

        protected override void RegisterOutputParams(GH_OutputParamManager pManager)
        {
            pManager.AddTextParameter("Fillet_JSON", "Fillet_JSON", "The compiled JSON string to feed into BIM_JSON.", GH_ParamAccess.item);
            pManager.AddTextParameter("Instructions", "Instructions", "Detailed usage instructions.", GH_ParamAccess.item);
        }

        protected override void SolveInstance(IGH_DataAccess DA)
        {
            double defaultRad = 0.0;
            DA.GetData(0, ref defaultRad);

            List<string> ruleTypes = new List<string>();
            DA.GetDataList(1, ruleTypes);

            List<string> ruleMatches = new List<string>();
            DA.GetDataList(2, ruleMatches);

            List<double> ruleRadii = new List<double>();
            DA.GetDataList(3, ruleRadii);

            List<bool> exactMatches = new List<bool>();
            DA.GetDataList(4, exactMatches);

            int ruleCount = Math.Min(Math.Min(ruleTypes.Count, ruleMatches.Count), ruleRadii.Count);
            
            var rules = new List<object>();

            for (int i = 0; i < ruleCount; i++)
            {
                bool isExact = true;
                if (i < exactMatches.Count)
                {
                    isExact = exactMatches[i];
                }

                rules.Add(new
                {
                    type = ruleTypes[i]?.Trim() ?? "",
                    match = ruleMatches[i]?.Trim() ?? "",
                    radius = ruleRadii[i],
                    exact = isExact
                });
            }

            var configDict = new
            {
                default_radius = defaultRad,
                rules = rules
            };

            string jsonOutput = JsonConvert.SerializeObject(configDict, Formatting.Indented);

            Message = $"{this.NickName}\nRULES BUILT\n---\nValid Rules: {ruleCount}\nDefault: {defaultRad}m";

            DA.SetData(0, jsonOutput);

            string instructions = @"FILLET RULE CONFIGURATOR
================================================================================
A UI helper node that compiles standard Grasshopper lists into a strict JSON 
schema for parametric corner rounding. Prevents syntax errors from manual typing.

HOW TO USE NEGATIVE RULES (e.g., ""Everything except Office""):
Because the Engine reads top-down, put your exception first, then a wildcard:
1. Type: Program | Match: Office | Radius: 0.0 (Exception)
2. Type: Program | Match: * | Radius: 3.0 (Everything Else)

INPUTS:
    DefaultRadius (float) : The base radius if no rules match.
    RuleTypes     (str)   : List of target types ('Tower', 'Program', 'Building').
    RuleMatches   (str)   : List of target names ('Main_Tower', 'Retail', '*').
    RuleRadii     (float) : List of radii to apply to the matched targets.
    ExactMatch    (bool)  [List Access] : True = Exact, False = Contains. Defaults to True.

OUTPUTS:
    Fillet_JSON   (str)   : The compiled JSON string to feed into BIM_JSON.
    Instructions  (str)   : Detailed usage instructions.
================================================================================";
            
            DA.SetData(1, instructions);
        }

        protected override System.Drawing.Bitmap Icon
        {
            get
            {
                return IconLoader.Load("Fillet_Rules.png");
            }
        }

        public override GH_Exposure Exposure => GH_Exposure.primary;

        public override Guid ComponentGuid
        {
            get { return new Guid("b8c1c4f0-32a4-43e6-94e8-8a8f15d9cf7b"); }
        }
    }
}
